plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,138 @@
+{
+    private static final Logger log = Logger.get(RedshiftUnloadSplitSource.class);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
+
+    private final CompletableFuture<Boolean> resultSetFuture;
+    private final String unloadOutputPath;
+    private final TrinoFileSystem fileSystem;
+
+    private Optional<Location> manifestLocation;
+    private boolean finished;
+
+    public RedshiftUnloadSplitSource(ExecutorService executor, Connection connection, PreparedStatement statement, String unloadOutputPath, TrinoFileSystem fileSystem)
+    {
+        requireNonNull(executor, "executor is null");
+        requireNonNull(connection, "connection is null");
+        requireNonNull(statement, "statement is null");
+        this.unloadOutputPath = requireNonNull(unloadOutputPath, "unloadOutputPath is null");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        resultSetFuture = CompletableFuture.supplyAsync(() -> {
+            log.debug("Executing: %s", statement);
+            try {
+                // Exclusively set readOnly to false to avoid query failing with "ERROR: transaction is read-only".
+                connection.setReadOnly(false);
+                return statement.execute();
+            }
+            catch (SQLException e) {
+                if (e instanceof RedshiftException && e.getMessage().contains("The S3 bucket addressed by the query is in a different region from this cluster")) {
+                    throw new TrinoException(REDSHIFT_S3_CROSS_REGION_UNSUPPORTED, "Redshift cluster and S3 bucket in different regions is not supported", e);
+                }
+                throw new RuntimeException(e);
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        return resultSetFuture
+                .thenApply(_ -> {
+                    ConnectorSplitBatch connectorSplitBatch = new ConnectorSplitBatch(getUnloadedFilePaths().stream()
+                            .map(fileInfo -> (ConnectorSplit) new RedshiftUnloadSplit(fileInfo.path, fileInfo.size))
+                            .collect(toImmutableList()), true);
+                    finished = true;
+                    return connectorSplitBatch;
+                });
+    }
+
+    private List<FileInfo> getUnloadedFilePaths()
+    {
+        ImmutableList.Builder<FileInfo> unloadedFilePaths = ImmutableList.builder();
+        manifestLocation = Optional.of(Location.of(unloadOutputPath + "manifest"));
+        TrinoInputFile inputFile = fileSystem.newInputFile(manifestLocation.get());
+        JsonNode outputFileEntries;
+        try (TrinoInputStream inputStream = inputFile.newStream()) {
+            byte[] manifestContent = inputStream.readAllBytes();
+            outputFileEntries = OBJECT_MAPPER.readTree(manifestContent).path("entries");
+        }
+        catch (FileNotFoundException e) {
```

Comment:
Why is this logic exception-driven instead of `inputFile.exists()`?

Reply:
@raunaqmorarka suggested to make it exception driven, possibly to avoid having to make an extra call to check if file exists. https://github.com/trinodb/trino/pull/24117#discussion_r1882011805

Reply:
Please leave a code comment.

Reply:
added